languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py

```diff
@@ -51,62 +52,100 @@ def _before_compile(query):
+    """Authorize an existing query with an Oso instance, user, and checked
+    permissions."""
     # Get the query session.
     session = query.session
 
-    # Check whether this is an oso session.
+    # Check whether this is an Oso session.
     if not isinstance(session, AuthorizedSessionBase):
         # Not an authorized session.
         return None
 
-    oso = session.oso_context["oso"]
+    oso: Oso = session.oso_context["oso"]
     user = session.oso_context["user"]
-    action = session.oso_context["action"]
+    checked_permissions: Permissions = session.oso_context["checked_permissions"]
+
+    # Early return if no authorization is to be applied.
+    if checked_permissions is None:
+        return None
 
     # TODO (dhatch): This is necessary to allow ``authorize_query`` to work
     # on queries that have already been made.  If a query has a LIMIT or OFFSET
     # applied, SQLAlchemy will by default throw an error if filters are applied.
     # This prevents these errors from occuring, but could result in some
     # incorrect queries. We should remove this if possible.
-    query = query.enable_assertions(False)
+    query = query.enable_assertions(False)  # type: ignore
 
     entities = {column["entity"] for column in query.column_descriptions}
     for entity in entities:
         # Only apply authorization to columns that represent a mapper entity.
         if entity is None:
             continue
 
-        authorized_filter = authorize_model(oso, user, action, query.session, entity)
+        # If entity is an alias, retrieve the underlying class.
+        alias = inspect(entity).class_ if isinstance(entity, AliasedClass) else None
+
+        # Only apply authorization to columns that have been specified as
+        # requiring authorization.
+        if alias in checked_permissions:
+            action = checked_permissions[alias]  # type: ignore
+        elif entity in checked_permissions:
+            action = checked_permissions[entity]  # type: ignore
+        else:
+            continue
+
+        session = query.session
+        assert isinstance(session, Session)
+        authorized_filter = authorize_model(oso, user, action, session, entity)
         if authorized_filter is not None:
-            query = query.filter(authorized_filter)
+            query = query.filter(authorized_filter)  # type: ignore
 
     return query
 
 
-def authorized_sessionmaker(get_oso, get_user, get_action, class_=None, **kwargs):
-    """Session factory for sessions with oso authorization applied.
-
-    :param get_oso: Callable that return oso instance to use for authorization.
-    :param get_user: Callable that returns user for an authorization request.
-    :param get_action: Callable that returns action for the authorization request.
+def authorized_sessionmaker(
+    get_oso: Callable[[], Oso],
+    get_user: Callable[[], Any],
+    get_checked_permissions: Callable[[], Permissions],
+    class_: Type[Session] = None,
```

Comment:
I added types to most of this file but haven't set up MyPy yet. Probably worth doing that & adding it to CI like some of the other Python libs.

Reply:
Apparently it's already checked in CI and was just waiting for us to add annotations.

Reply:
Yep! It's not set to require annotations.